### ISSUE TITLE
(Why not) support building larger-than-memory indexes

### ISSUE THREAD

**Issue by jbellis** [jbellis/jvector#168]
The DiskANN paper describes building larger-than-memory indexes by partitioning the dataset and then adding vectors to multiple partitions, then combining the graphs.  This is 2.5x slower than building the graph using brute force on a machine with a ton of memory.

I think a better approach would be to simply build the graph normally until we have `MAX_PQ_TRAINING_SET_SIZE` vectors added, then run the PQ algorithm and compress what we have so far.  When adding additional nodes, do PQ-based search with overquery and rerank, instead of comparing against raw vectors.

Tradeoffs:
1. You have to integrate writing to disk with graph construction, so you can get the original uncompressed vectors out of the heap.  MAYBE you end up dropping storing vectors inline with the graph edges.  (That would simplify writing to disk incrementally, and having the vectors inline doesn't seem like a huge performance win.)
2. You have to integrate deciding what kind of compression to use, with graph construction.
3. You give up ordinal renumbering.  However, in the specific case of C* compaction, rows will be added to the index in the "correct" order anyway, so no explicit renumbering would be needed anyway.

In short, I think it's doable, but it's also a pretty big change.

**Comment by xjtushilei** [comment]
I quite agree with the idea of reducing machine specifications through quantized compression. Looking forward to your big changes.
